### DISCUSSION TITLE
Adding libraries to cmake (QHull)

### DISCUSSION THREAD

**Discussion by eloirozier** [gnudatalanguage/gdl#962] [Q&A]
Hi everyone
I am currently working on adding Qhull to GDL (it seems to be a quite powerful and important library), but I'm having issues compiling and I can't even test my code.

Qhull headers are installed on my machine in /usr/include
I'm including headers from the libqhullcpp library in my cpp file like so:

> #include <libqhullcpp/RboxPoints.h>
> #include <libqhullcpp/QhullError.h>
> #include <libqhullcpp/QhullQh.h>
> #include <libqhullcpp/QhullFacet.h>

...

and then i use:

> using orgQhull::Qhull;
> using orgQhull::QhullError;

...

but then, even a simple line of code like declaring an object:

> Qhull qhull;

Makes the compilation crash at the end. This is what the terminal prompts:

> erozier@lp-mbroseta:~/GDL/GDL_GIT_erozier/build$ make -j 6
> [ 15%] Built target antlr
> Scanning dependencies of target gdl
> [ 16%] Building CXX object src/CMakeFiles/gdl.dir/libinit_jmg.cpp.o
> [ 16%] Building CXX object src/CMakeFiles/gdl.dir/qhull.cpp.o
> [ 16%] Linking CXX executable gdl
> /usr/bin/ld : CMakeFiles/gdl.dir/qhull.cpp.o : dans la fonction « lib::qhull(EnvT*) » :
> qhull.cpp:(.text+0xdd) : référence indéfinie vers « orgQhull::RboxPoints::RboxPoints() »
> /usr/bin/ld : qhull.cpp:(.text+0xe5) : référence indéfinie vers « orgQhull::Qhull::Qhull() »
> /usr/bin/ld : qhull.cpp:(.text+0xed) : référence indéfinie vers « orgQhull::Qhull::~Qhull() »
> /usr/bin/ld : qhull.cpp:(.text+0xf5) : référence indéfinie vers « orgQhull::RboxPoints::~RboxPoints() »
> /usr/bin/ld : CMakeFiles/gdl.dir/qhull.cpp.o : dans la fonction « lib::qhull(EnvT*) [clone .cold] » :
> qhull.cpp:(.text.unlikely+0x4) : référence indéfinie vers « orgQhull::RboxPoints::~RboxPoints() »
> collect2: error: ld returned 1 exit status
> make[2]: *** [src/CMakeFiles/gdl.dir/build.make:2426 : src/gdl] Erreur 1
> make[1]: *** [CMakeFiles/Makefile2:166 : src/CMakeFiles/gdl.dir/all] Erreur 2
> make: *** [Makefile:141 : all] Erreur 2

I modified the files CmakeLists.txt, config.h.cmake and CMakeModules.cmake/FindQHULL.cmake

It looks like (most probably) `cmake` or `make` does not take into account the libqhullcpp dependency at some step of the process...

But I am a complete noob with Cmake, and my additions seem to be unsifficient. Would you have any idea of what code is lacking in my workspace ? Thanks in advance for your insights, a solution would finally let me concentrate on the code itself

**Comment by GillesDuvert** [comment]
looks like the linker does not find the qhull shared library, libqhull_r.so . Others[ seems to have this problem](https://stackoverflow.com/questions/41592153/failure-to-link-qhull-c-interface-in-ros-catkin-project) and [here](https://stackoverflow.com/questions/61093084/how-to-use-qhull-as-an-external-project-in-cmake)
So the cmake did not do its job. 
Is the QHULL section of the CMakeLists.txt active (was commented out?)
Perhaps removing CMakeModules/FindQHULL.cmake  and do a cmake again, from scratch, will tell if QHULL is found (using generic cmake findqhull --- if any) and `-lqhullcpp -lqhull_r ` is added to the link command

**Reply by eloirozier** [reply]
I commented out the QHULL section in CMakeLists.txt
I removed the find_package() and put everything in CMakeLists.txt, like so:

```
if(QHULL)
  set(CMAKE_PREFIX_PATH ${QHULLDIR})
  # find_package(QHULL QUIET)
  # set(HAVE_QHULL ${QHULL_FOUND})
  set(HAVE_QHULL TRUE)
  set(QHULL_FOUND TRUE)

  if(QHULL_FOUND)
      
      find_path(
        QHULL_INCLUDE_DIR
        libqhullcpp/Qhull.cpp
        PATHS /usr/local/include /usr/include
      )

      find_library(
        QHULL_LIBRARY
        NAMES libqhullstatic libqhull qhullstatic qhull 
        PATHS /usr/lib64 /usr/local/lib /usr/lib
      )

      set(QHULL_LIBRARY_DIR "")
      get_filename_component(QHULL_LIBRARY_DIRS ${QHULL_LIBRARY} PATH)

      set(QHULL_INCLUDE_DIRS ${QHULL_INCLUDE_DIR})
      set(QHULL_LIBRARIES ${QHULL_LIBRARY})

      set(LIBRARIES ${LIBRARIES} ${QHULL_LIBRARIES})
      include_directories(${QHULL_INCLUDE_DIRS})

  else(QHULL_FOUND)
      message(FATAL_ERROR "QHULL is required but was not found.\n"
      "Use -DQHULLDIR=DIR to specify the QHULL directory tree.\n"
      "Use -DQHULL=OFF to not use it.\n"
                "(suitable Fedora/CentOS package libqhull-devel)")
  endif(QHULL_FOUND)
endif(QHULL)
```

And i get exactly the same result. I post this because there might be an obvious error here but i dont see it

Edit: this does not seem to work at all, i'm obviously doing smthg wrong

**Reply by GillesDuvert** [reply]
Would you do the reverse, let the QHULL section in CMakeLists.txt as it was, but force cmake to use its own qhull detection, not the one that was written by the GDL team (the CMakeModules/FindQHULL.cmake file, clearly obsolete) and see what happens? My impression is that QHULL has evolved a lot since somebody wrote the GDL FindQHULL.cmake. If cmake does not find qhull by itself, then certainly a up-to-date version of a FindQHULL.cmake can be found on the web. But the qhull part of CMakeLists.txt is certainly OK, I recall it:

```
# Qhull (enable TRIANGULATE and QHULL commands)
# -DQHULL=ON|OFF
# -DQHULLDIR=DIR
if(QHULL)
   set(CMAKE_PREFIX_PATH ${QHULLDIR})
   find_package(QHULL QUIET)
  set(PL_HAVE_QHULL ${QHULL_FOUND})
   if(QHULL_FOUND)
       set(LIBRARIES ${LIBRARIES} ${QHULL_LIBRARIES})
       include_directories(${QHULL_INCLUDE_DIR})
   else(QHULL_FOUND)
       message(FATAL_ERROR "QHULL is required but was not found.\n"
      "Use -DQHULLDIR=DIR to specify the QHULL directory tree.\n"
       "Use -DQHULL=OFF to not use it.\n"
                 "(suitable Fedora/CentOS package libqhull-devel)")
   endif(QHULL_FOUND)
 endif(QHULL)
```

**Reply by eloirozier** [reply]
Sorry i realize that i did not reply.
After searching for a recent example of a working FindQHULL.cmake, and doing some modifications i managed to make it work.

Briefly the problem was that i wanted to link the libqhullcpp library (qhull c++ interface), but i omitted to add beforehand the libqhull_r library on which it depends. Also the order of compilation was important (libqhull_r THEN libqhullpp)

In effect CMakeLists.txt was not to be changed.
Thanks for your help !
